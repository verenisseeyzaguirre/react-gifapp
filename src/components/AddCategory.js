import React from 'react'
import { useState } from 'react'
import PropTypes from 'prop-types'

export const AddCategory = ( { setCategories } ) => {

    const [inputValue, setInputValue] = useState(''); // si no habia '' era undefined y salia error

    const handleInputChange = ( e ) => {
        setInputValue( e.target.value )
        // console.log(e.target.value)
        console.log('handleInputChange llamado');
    }

    const handleSubmit = (e) => {
        e.preventDefault();
        console.log('handleSubmit llamado', inputValue)

        // cats es el estado anterior
        if ( inputValue.trim().length > 2 ) {
            setCategories( cats => [inputValue, ...cats] );
            setInputValue('');
        }
        
    }

    return (
        <form onSubmit= {handleSubmit}>
            <input 
                type="text"
                value = { inputValue }
                onChange = { handleInputChange }
                placeholder= 'Buscar gif...'
            />
        </form>
    )

}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
  };