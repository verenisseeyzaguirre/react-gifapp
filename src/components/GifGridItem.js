import React from 'react'


import PropTypes from 'prop-types';
//https://animate.style/

// param ( img ) en p {img.title}
export const GifGridItem = ( { title, url } ) => {

    //console.log( img );

    return (
        <div className='card animate__animated animate__fadeIn'>
            <img src={ url } alt={ title }/>
            <div className='name'>
                <p>{ title }</p>
            </div>
        </div>
    )
}

GifGridItem.propTypes = {
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired,
};

