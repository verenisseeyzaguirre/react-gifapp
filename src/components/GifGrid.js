import React from 'react'
//, { useState, useEffect }
import { GifGridItem } from './GifGridItem';
//import { getGifs } from '../helpers/getGifs';
import { useFetchGifs } from '../hooks/useFetchGifs';

export const GifGrid = ( { category } ) => {

    //const [count, setCount] = useState(0);
    //const [images, setImages] = useState([]);

    const { data:images, loading } = useFetchGifs( category );

    // [] indica q se ejecuando cuando se renderiza por primera vez y ya no mas
    // salio un warning porque si cambia la categoria deberia renderizar nuevmnte
    /*
    useEffect( () => {
        getGifs( category )
            .then( setImages );
    }, [ category ])
    */
    // getGifs()

    return (
        <>
            <h3 className='animate__animated animate__fadeIn' > {category} </h3>

            { loading && <p className='animate__animated animate__flash' >Loading</p> }

            <div className='card-grid' >
                {
                    images.map( img => (
                        <GifGridItem 
                            key={img.id}
                            {...img}
                        />
                    ))
                }         
            </div>
        </>
    )
}

/*
    debajo de h3
    <h2>{count}</h2>
    <button onClick= {()=> setCount(count+1) }>+1</button>
*/

/*
<ol>
    {
        images.map( img => (
            <li  key={img.id}>{img.title}</li>
        ))
    }
</ol>
*/

/*
{
    images.map( img => (
        <GifGridItem 
            key={img.id}
            {...img}
        />
    ))
}  
*/