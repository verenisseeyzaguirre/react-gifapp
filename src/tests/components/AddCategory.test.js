import React from 'react';
import '@testing-library/jest-dom';
//npm install -D @types/jest  importar para auto completado en los test

import { shallow } from 'enzyme';

import { AddCategory } from '../../components/AddCategory';

// { target : { value } }

describe('Pruebas en <AddCategory />', () => {

    //const setCategories = () => {};
    //no se sabe como es la fx de setCategories y por eso se usa la fx de abajo, tenemos la referencia como una fx
    // pensar en esa fx q es como la teniamos antes
    const setCategories = jest.fn();
    //const setInputValue = jest.fn();

    //se mantiene lo q sigue despues del igual porque si no ya no lee los atrib/metodos
    let wrapper = shallow( <AddCategory setCategories = {setCategories} /> );

    beforeEach( () => {
        //si teniamos una simulacion de algo , se limpie
        jest.clearAllMocks();
        wrapper = shallow( <AddCategory setCategories = {setCategories} /> );
    });

    test('Debe mostrar <AddCategory /> correctamente', () => {

        expect( wrapper ).toMatchSnapshot();

    });

    test('Debe de cambiar la caja de texto', () => {

        const input = wrapper.find('input');
        const value = 'Hola Mundo';
        input.simulate('change', { 
            target : {
                value : value
            }
        });
        const inputAfter = wrapper.find("input");
        expect(inputAfter.prop("value")).toBe(value);

    });

    test('No debe postear la informacion con submit', () => {
        //argumentos q seran transformados en el evento
        // metodo { prevenDefault: () => {} }
        wrapper.find("form").simulate('submit', { preventDefault(){} })

        expect( setCategories ).not.toHaveBeenCalled();

    });

    test('Debe llamar el setCategories y limpiar la caja del texto', () => {
        
        //1. Simular el inputChange
        const input = wrapper.find('input');
        const value = 'Hola Mundo';
        input.simulate('change', { 
            target : {
                value : value
            }
        });
        const inputAfter = wrapper.find("input");
        expect(inputAfter.prop("value")).toBe(value);

        //2. simular el submit
        wrapper.find("form").simulate('submit', { preventDefault(){} })
        //3. setCategories se debe haber llamado
        expect( setCategories ).toHaveBeenCalled();
        expect( setCategories ).toHaveBeenCalledTimes(1);
        //para confirmar q lo q ingresa a setCategories sea una funcion
        expect( setCategories ).toHaveBeenCalledWith( expect.any(Function) );
        //4.el valor del input debe ser ''
        const inputAfterTwice = wrapper.find("input");
        expect(inputAfterTwice.prop("value")).toBe('');
        //expect( setInputValue ).toHaveBeenCalled();

    });

})

/*
describe('Pruebas en <AddCategory/>', () => {
    let wrapper;
    const setCategories = jest.fn();
 
    const setInputValue = jest.fn();
    const useStateSpy = jest.spyOn(React, 'useState');
 
// ...
 
    test('debe de llamar el setInputValue', () => {
 
        useStateSpy.mockImplementation((init) => [init, setInputValue]);
 
        wrapper = shallow(<AddCategory setCategories={setCategories} />);
 
        const value = 'Hola Mundo';
        wrapper.find('input').simulate('change', {
            target: { value }
        });
        wrapper.find('form').simulate('submit', {
            preventDefault() { }
        })
 
        expect(setInputValue).toHaveBeenCalledWith('Hola Mundo');
        expect(setInputValue).toHaveBeenCalled();
 
    })
*/ 