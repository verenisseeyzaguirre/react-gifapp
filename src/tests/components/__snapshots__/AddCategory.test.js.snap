// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <AddCategory /> Debe mostrar <AddCategory /> correctamente 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    onChange={[Function]}
    placeholder="Buscar gif..."
    type="text"
    value=""
  />
</form>
`;
